### ISSUE TITLE
Explanation of Box in documentation is confusing

### ISSUE THREAD

**Issue by vchekan** [rust-lang/rust#39666]
In documentation on Box type
https://doc.rust-lang.org/book/choosing-your-guarantees.html#boxt

> Here, the box was moved into y. As x no longer owns it, the compiler will no longer allow the programmer to use x after this. A box can similarly be moved out of a function by returning it.

Any object, not just a box will be destructed when out of scope. Any moved binding will not be allowed to be used. What Box has to do with it?

Last sentence is even more puzzling. What is "it", a function? Then sentence makes no sense "moved out of function by returning function". If "it" is not a function, than what?

> This is a zero-cost abstraction for dynamic allocation. 

Wait, what does it have to do with dynamic allocation? This comes out of the blue without any logical introduction.

> If you want to allocate some memory on the heap and safely pass around a pointer to that memory, this is ideal

Again, no explanation, just some puzzling claims. where are pointers, how and why allocate memory, why pointer is safe now, how possibly pointer can be safe without storing size of allocated memory?

Bottom line is, after reading Box section I have no more understanding of Box than I had before reading it. After looking at source code, I start suspecting that `let x = Box::new(1);`will allocate "1" in heap, but there is nothing in the documentation which would say it explicitly. If so, what exactly will happen in given example, memory allocate for "1" and copy value, or just store pointer to existing "1" value?. Also I see "box" operator in Box implementation, what this is?

**Comment by steveklabnik** [comment]
We are not modifying the book anymore; we're working on a second edition, which has totally re-written all of this. As such, I'm going to give this issue a close; if you or anyone else wants to send in a PR, please do!
